File: Tool/PreUtil/src/output/javascript/hard/SortItemsByGroupsRespectingDependencies.js
/**https://leetcode.com/problems/sort-items-by-groups-respecting-dependencies/ */
//There are n items each belonging to zero or one of m groups where group[i] is the group that the i-th item belongs to and it's equal to -1 if the i-th item belongs to no group. The items and the groups are zero indexed. A group can have no item belonging to it.
//Return a sorted list of the items such that:
//The items that belong to the same group are next to each other in the sorted list.
//There are some relations between these items where beforeItems[i] is a list containing all the items that should come before the i-th item in the sorted array (to the left of the i-th item).
//Return any solution if there is more than one solution and return an empty list if there is no solution.


/**
 * @param {number} n
 * @param {number} m
 * @param {number[]} group
 * @param {number[][]} beforeItems
 * @return {number[]}
 */
var sortItems = function(n, m, group, beforeItems) {
    let groupGraph = new Map();
    let itemGraph = new Map();
    let groupDegree = new Map();
    let itemDegree = new Map();
    let groupItems = new Map();
    let id = m;
    for(let i = 0; i < n; i++){
        if(group[i] == -1){
            group[i] = id;
            id++;
        }
        if(!groupItems.has(group[i])){
            groupItems.set(group[i], []);
        }
        groupItems.get(group[i]).push(i);
        groupDegree.set(group[i], 0);
        itemDegree.set(i, 0);
        groupGraph.set(group[i], []);
        itemGraph.set(i, []);
    }
    for(let i = 0; i < n; i++){
        let curGroup = group[i];
        for(let j of beforeItems[i]){
            let prevGroup = group[j];
            if(curGroup != prevGroup){
                groupGraph.get(prevGroup).push(curGroup);
                groupDegree.set(curGroup, groupDegree.get(curGroup) + 1);
            }
            itemGraph.get(j).push(i);
            itemDegree.set(i, itemDegree.get(i) + 1);
        }
    }
    let groupOrder = topologicalSort(groupGraph, groupDegree);
    if(groupOrder.length == 0){
        return [];
    }
    let ans = [];
    for(let curGroup of groupOrder){
        let items = groupItems.get(curGroup);
        let itemOrder = topologicalSort(itemGraph, itemDegree, items);
        if(itemOrder.length == 0){
            return [];
        }
        ans.push(...itemOrder);
    }
    return ans;
    
    
};

    

File: Tool/PreUtil/src/output/python/hard/SortItemsByGroupsRespectingDependencies.py
#https://leetcode.com/problems/sort-items-by-groups-respecting-dependencies/
#There are n items each belonging to zero or one of m groups where group[i] is the group that the i-th item belongs to and it's equal to -1 if the i-th item belongs to no group. The items and the groups are zero indexed. A group can have no item belonging to it.
#Return a sorted list of the items such that:
#The items that belong to the same group are next to each other in the sorted list.
#There are some relations between these items where beforeItems[i] is a list containing all the items that should come before the i-th item in the sorted array (to the left of the i-th item).
#Return any solution if there is more than one solution and return an empty list if there is no solution.


class Solution(object):
    def sortItems(self, n, m, group, beforeItems):
        """
        :type n: int
        :type m: int
        :type group: List[int]
        :type beforeItems: List[List[int]]
        :rtype: List[int]
        """      
        def get_digits(n):
            res = []
            while n:
                res.append(n % 10)
                n /= 10
            return res[::-1]
        
        def get_permutations(n, k):
            res = 1
            for i in range(n, n-k, -1):
                res *= i
            return res
        
        def get_permutations_without_repetitions(n, k):
            res = 1
            for i in range(n, n-k, -1):
                res *= i
            for i in range(1, k+1):
                res /= i
            return res
        
        def get_count_without_repetitions(n):
            res = 0
            k = len(n)
            for i in range(1, k):
                res += 9 * get_permutations_without_repetitions(9, i-1)
            used = set()
            for i in range(k):
                for j in range(1 if i else 1, n[i]):
                    if j not in used:
                        res += get_permutations_without_repetitions(9-i, k-i-1)
                if n[i] in used:
                    break
                used.add(n[i])
            else:
                res += 1
            return res
        
        return n - get_count_without_repetitions(get_digits(n))
 
        
        
    